File: archive/r/rust/insertion-sort.rs
use std::env::args;
use std::process::exit;
use std::str::FromStr;

fn usage() -> ! {
    println!("Usage: please provide a list of at least two integers to sort in the format \"1, 2, 3, 4, 5\"");
    exit(0);
}

fn parse_int<T: FromStr>(s: &str) -> Result<T, <T as FromStr>::Err> {
    s.trim().parse::<T>()
}

fn parse_int_list<T: FromStr>(s: &str) -> Result<Vec<T>, <T as FromStr>::Err> {
    s.split(',')
        .map(parse_int)
        .collect::<Result<Vec<T>, <T as FromStr>::Err>>()
}

fn insertion_sort<T: PartialOrd>(arr: &mut Vec<T>) {
    for i in 0..arr.len() {
        for j in (0..i).rev() {
            if arr[j] >= arr[j+1] {
                arr.swap(j, j+1);
            } else {
                break;
            }
        }
    }
}

fn main() {
    let mut args = args().skip(1);

    // Convert 1st command-line argument to list of integers
    let mut arr: Vec<i32> = args
        .next()
        .and_then(|s| parse_int_list(&s).ok())
        .unwrap_or_else(|| usage());

    // Exit if list too small
    if arr.len() < 2 {
        usage();
    }

    insertion_sort(&mut arr);
    println!("{arr:?}");
}

File: archive/r/ruby/insertion-sort.rb
# frozen_string_literal: true

class Array
  def insertion_sort!
    (1...length).each do |i|
      value = self[i]
      j = i - 1

      while j >= 0 && self[j] > value
        self[j + 1] = self[j]
        j -= 1
      end

      self[j + 1] = value
    end

    self
  end
end

def usage!
  abort %(Usage: please provide a list of at least two integers to sort in the format "1, 2, 3, 4, 5")
end

def parse_input
  raw = ARGV.first
  usage! if raw.nil? || raw.strip.empty?

  numbers = raw.split(",").map { Integer(it.strip) }
  usage! if numbers.length < 2

  numbers
rescue ArgumentError
  usage!
end

puts parse_input.insertion_sort!.join(", ")
